packages/next-server/lib/fetch/retry.ts

```diff
@@ -0,0 +1,77 @@
+export type RetryOptions = {
+  retries?: number
+  factor?: number
+  minTimeout?: number
+  onRetry?(err: Error, attempt: number): void,
+}
+
+export class FetchError extends Error {
+  readonly statusCode: number
+  readonly status: number
+  readonly code: number
```

Comment:
What is the reason to have all 3?

Reply:
`code` is very common in error objects, `status` is used by `res` and users may try that one, and `statusCode` is also very used to get the status code, specially in implementations where `code` and `statusCode` may mean something different.

Reply:
[`code` in `Error` objects](https://nodejs.org/api/errors.html#errors_error_code) is expected to be a `string`, not a `number`. and something like `'ERR_HTTP_STATUS_CODE'` probably makes more sense than putting the exact HTTP code.

Furthermore, having two different properties denoting the same value is a bit strange. If you provide proper typings and people use a reasonable advanced editor (with proper autocomplete), none of this is necessary. Also, this way you're forcing all downstream code to provide the same behavior if they want to be consistent.

Reply:
@Janpot I'm open to ideas, what do you suggest ?, The current implementation of `FetchError` is only used for the case of failed fetch requests, nothing else, there's only a status code and message

Reply:
If I think of how I usually need to consume these types of errors, I'd say it would be like
```js
async function fetchMyThing () {
  try {
    return await fetch('/my/thing');
  } catch (err) {
    if (err.code === 'ERR_HTTP_STATUS_CODE') {
      // there may be other classes of errors thrown than status code errors,
      // so we filtered them out first.
      if (err.status === 404) {
        return null;
      } else if (err.status === 401) {
        await requestLogin();
        return fetchMyThing();
      }
    }
    throw err;
  }
}
```

Reply:
@Janpot I'm 100% into adding `code` as a `string`, the actual `code`  to add can be discussed, we can add something like `FETCH_ERROR` for all request and let the users handle the error using `status`, or add a different code depending on the `status`, like `NOT_FOUND` if `404` , although I don't know how useful can that be.

keeping `status` and `statusCode` with the same value can be better then, the idea is to have an isomorphic fetch, it should work in the same way for both sides.